tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py

```diff
@@ -0,0 +1,356 @@
+    image = x["image"].to(device)
+
+    with torch.no_grad():
+        image_recon = autoencoderkl.reconstruct(image)
+
+    mmd_scores.append(mmd(image, image_recon))
+
+mmd_scores = torch.stack(mmd_scores)
+print(f"MS-SSIM score: {mmd_scores.mean().item():.4f} +- {mmd_scores.std().item():.4f}")
+
+# -
+
+# # Compute MultiScaleSSIMMetric and SSIMMetric
+#
+# Both MS-SSIM and SSIM can be used as metric to evaluate the diversity.
+#
+# SSIM measures the similarity between two images based on three components: luminance, contrast, and structure. In addition, MS-SSIM is an extension of SSIM that computes the structural similarity measure at multiple scales. Both metrics can assume values between 0 and 1, where 1 indicates perfect similarity between the images.
+#
+# In this section we will compute the MS-SSIM and SSIM Meteric between the real images and those reconstructed by the AutoencoderKL.
+
+# +
+ms_ssim_recon_scores = []
+ssim_recon_scores = []
+autoencoderkl.eval()
+
+ms_ssim = MultiScaleSSIMMetric(spatial_dims=2, data_range=1.0, kernel_size=4)
+ssim = SSIMMetric(spatial_dims=2, data_range=1.0, kernel_size=4)
+
+for step, x in list(enumerate(val_loader)):
+    image = x["image"].to(device)
+
+    with torch.no_grad():
+        image_recon = autoencoderkl.reconstruct(image)
+
+    ms_ssim_recon_scores.append(ms_ssim(image, image_recon))
+    ssim_recon_scores.append(ssim(image, image_recon))
+
+ms_ssim_recon_scores = torch.cat(ms_ssim_recon_scores, dim=0)
+ssim_recon_scores = torch.cat(ssim_recon_scores, dim=0)
+
+print(f"MS-SSIM Metric: {ms_ssim_recon_scores.mean():.7f} +- {ms_ssim_recon_scores.std():.7f}")
+print(f"SSIM Metric: {ssim_recon_scores.mean():.7f} +- {ssim_recon_scores.std():.7f}")
+
+# -
+
+# Compute the SSIM and MS-SSIM between synthetic and real images. Note that here we are regenerating some synthetic images.
+
+# +
+ms_ssim_scores = []
+ssim_scores = []
+unet.eval()
+
+for step, x in enumerate(val_loader):
+    # Get the real images
+    real_images = x["image"].to(device)
+
+    # Generate some synthetic images using the defined model
+    n_synthetic_images = len(x['image'])
+    noise = torch.randn((n_synthetic_images, 1, 64, 64))
+    noise = noise.to(device)
+    scheduler.set_timesteps(num_inference_steps=25)
+
+    with torch.no_grad():
+        syn_images, intermediates = inferer.sample(input_noise=noise, diffusion_model=unet,
+                                                  scheduler=scheduler,save_intermediates=True,
+                                                  intermediate_steps=100)
+
+        ms_ssim_scores.append(ms_ssim(real_images, syn_images))
+        ssim_scores.append(ssim(real_images, syn_images))
+
+ms_ssim_scores = torch.cat(ms_ssim_scores, dim=0)
+ssim_scores = torch.cat(ssim_scores, dim=0)
+
+print(f"MS-SSIM Metric: {ms_ssim_scores.mean():.7f} +- {ms_ssim_scores.std():.7f}")
+print(f"SSIM Metric: {ssim_scores.mean():.7f} +- {ssim_scores.std():.7f}")
```

Comment:
to measure diversity, it is necessary to compute the score between synthetic and synthetic images.